docs/debug-adapter.md

```diff
@@ -0,0 +1,37 @@
+---
+id: dap-integration
+title: Debug Adapter Protocol Integration 
+sidebar_label: DAP Integration guide
+---
+
+## Debug Adapter Protocol integration guide
+
+Bloop allows the client to run and debug both the main classes and the tests defined in the project. 
+This is possible thanks to the Debug Adapter Protocol, which bloop exposes to the client 
+in a run/test task start notification. This adapter expects either a 
+`ScalaTestClassesItem` or `ScalaMainClass` as a configuration parameter for the launch request.
+
+### Client-Server communication
+![communication sequence diagram](assets/dap-sequence.png)
+
+1. BSP Client chooses which main class to run. To do so, it requests the list of all runnable main classes from the BSP Server.
+2. BSP Client request the chosen class to be run.
```

Comment:
@jvican written:
> I really like that we use `build/taskStart` as the notification to tell the BSP client the connection details. Another alternative would be to use a custom Bloop-specific notification.

@olafurpg written:
> I like the idea of keeping `buildTarget/run` and establish the debug adapter session via `build/taskStart` notifications

Reply:
Since the `run` request responds with the report for the whole run, using a notification it is the only viable option. And I think that creating another, custom notification would only add to the complexity. 

I have one concern: I have proposed to use the `RunTask` as a data field in`build/taskStart` - but what about tests? I am not sure, but aren't these notifications sent for each suite/build target? 
That's why, I would now like to propose sending `build/taskStart` with `DebugAdapterTask` as data instead. Then both the `RunTask` and `TestTask` would remain unchanged.

Reply:
> I am no longer sure if it's a good idea to establish DAP via buildTarget/run
I think it it would be best to introduce a custom JSON-RPC endpoint to start a debug session

Why? What are your concerns?

Reply:
I think it complicates the pipeline to keep `buildTarget/run` and `buildTarget/test` in the mix. The two endpoints can be replaced with `debugLaunchConfiguration/resolve` and it would be clearer for a new person to read the spec and understand what's going on.

Reply:
In that case:
> The two endpoints can be replaced with

Fully agree

> `debugLaunchConfiguration/resolve`

we could work on the name, since the it is a bit misleading (and a mouthful): DAP is just an implementation detail. 
First part could be something related to launching, starting, evaluating? 
Second could be any of: `startSession`, `openConnection`, and probably more?

Reply:
To clarify, I'm not proposing to remove `buildTarget/run` and `buildTarget/test` from BSP. Those two endpoints can be dealt with separately. My primary suggestion is to not bake the Debug Adapter Protocol into the semantics of `buildTarget/run` and `buildTarget/test`. I think it's clearer to have a custom endpoint `debugLaunchConfiguration/resolve` that deals only with how to establish a DAP session via LSP/BSP.

Reply:
My thinking behind the name of `debugLaunchConfiguration/resolve`:

- DAP is not an implementation detail since the contract of the `port` field is that the client will be able to open a socket to that port number to speak DAP.
- it resembles `completionItem/resolve` from LSP, which is used in a similar way: client sends a `DebugLaunchConfiguration` JSON request and the response contains a `DebugLaunchConfiguration` JSON response with all fields populated.

Reply:
So, as I understand it: you propose that we:
- introduce a new endpoint for dealing with the DAP
- leave the `buildTarget/run/test` as they were

So now, when the user wants to run the main class, it uses the `buildTarget/run` endpoint but when he wants to debug it, he has to spin up a Debug Client and run `debugLaunch.../resolve`.
If that is not correct, please elaborate on your solution, because I am clearly missing something :(

Reply:
> So now, when the user wants to run the main class, it uses the buildTarget/run endpoint

This is incorrect, the user always runs `debugLaunchConfiguration/resolve` (both for main/test) and always needs to spin up a Debug Client. We can rename the endpoint to `launchConfiguration/resolve` if that makes it clearer. The `buildTarget/run` and `buildTarget/test` can be left unchanged in the current BSP spec but Metals won't use them.

Reply:
I am asking from the viewpoint of a BSP Client. If we leave "dangling endpoints", someone is bound to use them. If you consider this a non-issue, please say so.

Also, because I might have the wrong idea the whole time - are we brainstorming integrating the `DAP` with `bloop`, or are we spiking a marriage of `DAP` and `BSP`?

Reply:
I think it's a non-issue to leave `buildTarget/run` and `buildTarget/test` unchanged for now. There may exist BSP clients which appreciate a simpler interface to run/test without the complexity of the DAP.

> marriage of DAP and BSP

We're integrating DAP into BSP by introducing a new request endpoint `launchConfiguration/resolve` that enables the build server to power a rich user interface for launching/evaluating source code with optional debugging functionality. I wouldn't call it a marriage since the two protocols can live fine without each other.

Reply:
DAP has a lot of nice functionality and rich client/server implementations that we have no desire to merge in BSP, and BSP has a lot of functionality and rich client/server implementations that we have no desire to merge into DAP.

Reply:
As discussed offline, the new JSON-RPC endpoint would be

```ts
// request: 'dap/connection' or 'debugAdapterProtocol/connection'
// params: empty object {}
// response: DebugAdapterProtocolConnection
interface DebugAdapterProtocolConnection {
  // Port number for client to establish a DAP session.
  port: number;
}
```

The `launchData` field has been removed, it will be the responsibility of Metals to figure out how to resolve the launch arguments.